pkg/collector/corechecks/snmp/gosnmp_value.go

```diff
@@ -0,0 +1,154 @@
+			continue
+		}
+		returnValues[name] = value
+	}
+	return returnValues
+}
+
+// resultToColumnValues builds column values
+// - columnResultValuesType: column values
+// - nextOidsMap: represent the oids that can be used to retrieve following rows/values
+func resultToColumnValues(columnOids []string, snmpPacket *gosnmp.SnmpPacket) (columnResultValuesType, map[string]string) {
+	returnValues := make(columnResultValuesType, len(columnOids))
+	nextOidsMap := make(map[string]string, len(columnOids))
+	maxRowsPerCol := int(math.Ceil(float64(len(snmpPacket.Variables)) / float64(len(columnOids))))
+	for i, pduVariable := range snmpPacket.Variables {
+		if shouldSkip(pduVariable.Type) {
+			continue
+		}
+
+		oid, value, err := getValueFromPDU(pduVariable)
+		if err != nil {
+			log.Debugf("Cannot get value for variable `%v` with type `%v` and value `%v`", pduVariable.Name, pduVariable.Type, pduVariable.Value)
+			continue
+		}
+		// the snmpPacket might contain multiple row values for a single column
+		// and the columnOid can be derived from the index of the PDU variable.
+		columnOid := columnOids[i%len(columnOids)]
+		if _, ok := returnValues[columnOid]; !ok {
+			returnValues[columnOid] = make(map[string]snmpValueType, maxRowsPerCol)
+		}
+
+		prefix := columnOid + "."
+		if strings.HasPrefix(oid, prefix) {
+			returnValues[columnOid][oid[len(prefix):]] = value
+			nextOidsMap[columnOid] = oid
+		} else {
+			// if oid is not prefixed by columnOid, it means it's not part of the column
+			// and we can stop requesting the next row of this column
+			delete(nextOidsMap, columnOid)
+		}
+	}
+	return returnValues, nextOidsMap
+}
+
+func shouldSkip(berType gosnmp.Asn1BER) bool {
+	switch berType {
+	case gosnmp.EndOfContents, gosnmp.EndOfMibView, gosnmp.NoSuchInstance, gosnmp.NoSuchObject:
+		return true
+	}
+	return false
+}
+
+// getSubmissionType converts gosnmp.Asn1BER type to submission type
+//
+// ZeroBasedCounter64: We don't handle ZeroBasedCounter64 since it's not a type currently provided by gosnmp.
+// This type is currently supported by python impl: https://github.com/DataDog/integrations-core/blob/d6add1dfcd99c3610f45390b8d4cd97390af1f69/snmp/datadog_checks/snmp/pysnmp_inspect.py#L37-L38
+func getSubmissionType(gosnmpType gosnmp.Asn1BER) string {
+	switch gosnmpType {
+	// Counter Types: From the snmp doc: The Counter32 type represents a non-negative integer which monotonically increases until it reaches a maximum
+	// value of 2^32-1 (4294967295 decimal), when it wraps around and starts increasing again from zero.
+	// We convert snmp counters by default to `rate` submission type, but sometimes `monotonic_count` might be more appropriate.
+	// To achieve that, we can use `forced_type: monotonic_count` or `forced_type: monotonic_count_and_rate`.
+	case gosnmp.Counter32, gosnmp.Counter64:
+		return "counter"
+	}
+	return ""
```

Comment:
what about `Integer` and `UInteger` ?

Reply:
I'm not sure to understand. You mean using constants with `int/uint` value ?

Note this is used here: https://github.com/DataDog/datadog-agent/blob/2c56b27a18aa3370871b887f9b3a8d57dc2e8707/pkg/collector/corechecks/snmp/report_metrics.go#L81

Reply:
I would expect the `getSubmissionType` to return `gauge` as default for `gosnmp.Integer` and `gosnmp.UInteger`, rather than `""`. If the function only returns a value for `Counter` it should be called `getCounterSubmissionType` and could even be a dictionary vs a function

Reply:
One reason I left the default `gauge` to later stage [here in `sendMetric`](https://github.com/DataDog/datadog-agent/blob/c582e592c970ca35d946a918f04896e97b55d655/pkg/collector/corechecks/snmp/report_metrics.go#L81-L85) is that `getSubmissionType()` is called for all values, including string values for tags, and setting "gauge" might be semantically misleading if the value is a string used for tags.

Both way are possible, I don't have strong opinion on it. I can change to return "gauge" if you prefer :)